tsdb/engine/tsm1/engine.go

```diff
@@ -700,111 +698,87 @@ func (e *Engine) Export(w io.Writer, basePath string, start time.Time, end time.
 			tombstonePath = filepath.Base(r.TombstoneFiles()[0].Path)
 		}
 
-		// Skip this file if no overlap with start/end
 		min, max := r.TimeRange()
-		if min > end.UnixNano() || max < start.UnixNano() {
-			r.Close()
-			continue
-		}
 
-		r.Close()
+		// We overlap time ranges, we need to filter the file
+		if min >= start.UnixNano() && min <= end.UnixNano() ||
+			max >= start.UnixNano() && max <= end.UnixNano() {
+			if err := e.filterFileToGzip(r, file, basePath, filepath.Join(path, file), start.UnixNano(), end.UnixNano(), zw); err != nil {
+				return err
+			}
 
-		if tombstonePath != "" {
-			whole = append(filtered, tombstonePath)
 		}
 
-		if min > start.UnixNano() && max < end.UnixNano() {
-			whole = append(filtered, file)
-		} else {
-			// at this point, we know we're not totally outside or inside start and end so we'll need to scan the blocks
-			filtered = append(filtered, file)
-		}
-	}
+		// above is the only case where we need to keep the reader open.
+		r.Close()
 
-	for _, f := range filtered {
-		if err := e.filterFileToBackup(f, basePath, filepath.Join(path, f), start.UnixNano(), end.UnixNano(), tw); err != nil {
-			return err
+		// the TSM file is 100% inside the range, so we can just write it without scanning each block
+		if min > start.UnixNano() && max < end.UnixNano() {
+			if err := e.writeFileToGzip(file, basePath, filepath.Join(path, file), zw); err != nil {
+				return err
+			}
 		}
-	}
 
-	for _, f := range whole {
-		if err := e.writeFileToBackup(f, basePath, filepath.Join(path, f), tw); err != nil {
-			return err
+		// if this TSM file had a tombstone we'll write out the whole thing too.
+		if tombstonePath != "" {
+			if err := e.writeFileToGzip(tombstonePath, basePath, filepath.Join(path, tombstonePath), zw); err != nil {
+				return err
+			}
 		}
+		// when writing a file to Gzip, zw.Close() is how you signify you are done writing the file.
+		// so the helper functions will do the closing.
+		// to write the next file to the gzip stream, we just reset the gzip writer.
+		// later, this zipped stream must be read with zr.Multistream(false) so that it will stop reading on each new file.
```

Comment:
I don't think we use multistream on import.  This may break some existing things in enterprise.